8bit Optimisers and resolved potential setup.py issues

korakoe opened the pull request:
The commit to setup.py was to resolve dependency issues on my machine, as certain features are deprecated in later versions of lightning and torchmetrics, these versions still have these features but present deprecation warnings, feel free to change these as you see necessary

Also, regarding 8bit, replacing `nn.Linear()` with its 8bit equivalent (`bnb.nn.Linear8bitLt()` | which seems to be only for inference) seems to break tensors, so I've opted for merely implementing the optimisers

korakoe committed:
8bit Optimisers and resolved potential Env issues

The commit to setup.py was to resolve dependency issues on my machine, as certain features are deprecated in later versions of lightning and torchmetrics, these versions still have these features but present deprecation warnings

Also, regarding 8bit, replacing nn.Linear() with it's 8bit equivalent seems to break tensors, so I've opted for merely implementing the optimisers

korakoe committed:
grammatical correctness

korakoe committed:
Forgot weight decay

isamu-isozaki commented:
Looks good! Sorry, slight merge conflicts but once resolved will merge

isamu-isozaki commented:
Also, for 
```
import bitsandbytes as bnb
```
can you do some error checking? In some systems like mine, this causes an exception even when it's installed like
```
CUDA Setup failed despite GPU being available.
```

korakoe commented:
> Also, for
> 
> ```
> import bitsandbytes as bnb
> ```
> 
> can you do some error checking? In some systems like mine, this causes an exception even when it's installed like
> 
> ```
> CUDA Setup failed despite GPU being available.
> ```

Sure, will do, sorry for the late reply

isamu-isozaki commented:
Thanks!

korakoe committed:
Add error checking for 8bit

unfortunately, bitsandbytes uses a broad exception clause, so I've decided to check whether "CUDA SETUP" is contained within the error message, if it is it defaults to a standard optimiser, if it isnt it raises the error

not the most elegant solution but until bitsandbytes adds proper errors its pretty hard to filter these out

korakoe commented:
> Thanks!

I've discovered this is a windows only issue, which requires some workarounds to fix, I've implemented error checking however for systems without these tweaks, I suggest reading the commit comment

if you'd prefer to just discard the optimiser under any error I can do that

isamu-isozaki commented:
@korakoe ah, can you do something like https://github.com/huggingface/diffusers/blob/main/examples/textual_inversion/textual_inversion.py for this? They mainly check if bitsandbytes can load and if not just discard it pretty much like you did but will be great if it's abstracted away into a function

korakoe commented:
> @korakoe ah, can you do something like https://github.com/huggingface/diffusers/blob/main/examples/textual_inversion/textual_inversion.py for this? They mainly check if bitsandbytes can load and if not just discard it pretty much like you did but will be great if it's abstracted away into a function

Sure, will do, although I can’t seem to find how they did it in this script, other scripts seem to use an import error instead and import it when it’s needed

isamu-isozaki commented:
@korakoe Ah, sorry. Yup, like the other scripts. Thanks a bunch

korakoe committed:
ImportError catching

korakoe committed:
Merge branch 'adding_training_script' into adding_training_script

korakoe commented:
> @korakoe Ah, sorry. Yup, like the other scripts. Thanks a bunch

Done, also resolved merge conflicts while I was at it
